drivers/memory_window/dynamic_slots.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "dynamic_slots.h"
+
+#if CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+#include <zephyr/drivers/mm/system_mm.h>
+#endif /* CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB */
+#include <zephyr/arch/xtensa/cache.h>
+
+#if CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+static uint8_t tlb[MEMORY_WINDOW_SLOTS_COUNT];
+#endif /* CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB */
+
+uint32_t dynamic_slots_init(const struct device *dev)
+{
+	uint32_t buf_size = MEMORY_WINDOW_SLOT_SIZE * MEMORY_WINDOW_SLOTS_COUNT;
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t *virt_buf = (uint32_t __sparse_cache *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+
+	memset(virt_buf, 0, (buf_size+MEMORY_WINDOW_SLOT_SIZE));
+
+	return 0;
+}
+
+uint32_t dynamic_slots_map_slot(uint32_t slot_id, const struct device *dev,
+		uint32_t resource_id, void *virtual_memory_address)
+{
+
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t *buf = (uint32_t __sparse_cache *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+	uint32_t buf_size =
+		MEMORY_WINDOW_SLOT_SIZE + MEMORY_WINDOW_SLOT_SIZE*MEMORY_WINDOW_SLOTS_COUNT;
+
+	z_xtensa_cache_flush_inv(buf, buf_size);
+
+	struct memory_window_struct *memory_window =
+			(struct memory_window_struct *)(z_soc_uncached_ptr(buf));
```

Comment:
here `buf` must be `__sparse_cache` or must be type-cast to it. So, maybe in this patch it would be good to declare `uint32_t __sparse_cache *buf`. But check where else it is used, maybe you'll end up casting it back to the default address space too often